src/lib/dialog/dialog-content-directives.ts

```diff
@@ -34,7 +42,19 @@ export class MatDialogClose implements OnChanges {
 
   @Input('matDialogClose') _matDialogClose: any;
 
-  constructor(public dialogRef: MatDialogRef<any>) { }
+  constructor(
+    @Optional() public dialogRef: MatDialogRef<any>,
+    private _elementRef: ElementRef,
+    private _dialog: MatDialog) {}
+
+  ngOnInit() {
+    if (!this.dialogRef) {
+      // If we couldn't get a dialog ref through DI, try to find one by looking at the DOM.
+      // This can happen if the close button is used inside a template dialog. Note that
+      // we have to do the lookup in OnInit, because the constructor might be too early.
```

Comment:
Small comment suggestion:
```ts
// When this directive is included in a dialog via TemplateRef (rather than being
// in a Component), the DialogRef isn't available via injection because embedded
// views cannot be given a custom injector. Instead, we look up the DialogRef by
// ID. This must occur in `onInit`, as the ID binding for the dialog container won't
// be resolved at constructor time.
```